### PR TITLE
Df mesh data read

### PR DESCRIPTION
**Pull request by dflanderka**

### CODE REVIEW

File: src/fields/field_fe.cc [hunk 1]
```diff
@@ -687,15 +682,17 @@ void FieldFE<spacedim, Value>::calculate_native_values(ElementDataCache<double>:
...
 	VectorMPI::VectorDataPtr data_vector = data_vec_.data_ptr();
 	std::vector<LongIdx> global_dof_indices(dh_->max_elem_dofs());
+	std::vector<int> const & el_ids = ReaderCache::get_reader(reader_file_)->get_element_vector(this->boundary_domain_);
 
 	// iterate through cells, assembly MPIVector
 	for (auto cell : dh_->own_range()) {
 		dof_size = cell.get_dof_indices(global_dof_indices);
 		LocDofVec loc_dofs = cell.get_loc_dof_indices();
-		data_vec_i = cell.elm_idx() * dof_size;
+		auto pos = std::find(el_ids.begin(), el_ids.end(), source_target_mesh_elm_map_[cell.elm_idx()]) - el_ids.begin();
```
Comment (jbrezmorf) [inline src/fields/field_fe.cc:691]: Change meaning of `source_target_mesh_elm_map_` it should map element index in the calculation mesh to the row index in the element cache

File: src/mesh/mesh.cc [hunk 2]
```diff
@@ -889,31 +889,32 @@ bool compare_points(const arma::vec3 &p1, const arma::vec3 &p2) {
...
         // to each node must be found just only one node in target \p mesh
         // store orders (mapping between source and target meshes) into node_ids vector
         std::vector<unsigned int> searched_elements; // for BIH tree
         unsigned int i_node, i_elm_node;
-        const BIHTree &bih_tree=mesh.get_bih_tree();
+        const BIHTree &bih_tree=this->get_bih_tree();
 
     	// create nodes of mesh
-        node_ids.resize( this->n_nodes() );
+        node_ids.resize( computational_mesh.n_nodes() );
```
Comment (jbrezmorf) [inline src/mesh/mesh.cc:910]: E.g. here just initialize to undef_idx and then pass through the input mesh, tr to find the node in the computation mesh and possibly mark the input node index in to the map.

File: src/mesh/mesh.cc [hunk 3]
```diff
@@ -889,31 +889,32 @@ bool compare_points(const arma::vec3 &p1, const arma::vec3 &p2) {
...
+	element_ids_map.resize(computational_mesh.n_elements()+computational_mesh.n_elements(true));
 
     {
         // iterates over node vector of \p this object
         // to each node must be found just only one node in target \p mesh
         // store orders (mapping between source and target meshes) into node_ids vector
         std::vector<unsigned int> searched_elements; // for BIH tree
         unsigned int i_node, i_elm_node;
-        const BIHTree &bih_tree=mesh.get_bih_tree();
+        const BIHTree &bih_tree=this->get_bih_tree();
```
Comment (jbrezmorf) [inline src/mesh/mesh.cc:907]: Rather make `this` the computational mesh and pass the input mesh in. Also use the BIH tree of the computational mesh as that is constructed only once.

File: src/fields/field_fe.cc [hunk 4]
```diff
@@ -311,27 +311,22 @@ void FieldFE<spacedim, Value>::set_mesh(const Mesh *mesh, bool boundary_domain)
...
 		this->boundary_domain_ = boundary_domain;
+		auto source_mesh = ReaderCache::get_mesh(reader_file_);
+		ReaderCache::get_element_ids(reader_file_, *(source_mesh.get()));
 		switch (this->interpolation_) {
 			case DataInterpolation::identic_msh:
-				ReaderCache::get_element_ids(reader_file_, *mesh);
 				break;
 			case DataInterpolation::equivalent_msh:
-				if (!ReaderCache::check_compatible_mesh( reader_file_, const_cast<Mesh &>(*mesh) )) {
+				if (!ReaderCache::get_mesh(reader_file_)->check_compatible_mesh(const_cast<Mesh &>(*mesh), source_target_mesh_elm_map_)) {
```
Comment (jbrezmorf) [inline src/fields/field_fe.cc:320]: Keep support for the identic mesh case.

File: src/fields/field_fe.cc [hunk 5]
```diff
@@ -440,7 +435,7 @@ bool FieldFE<spacedim, Value>::set_time(const TimeStep &time) {
 		bool is_native = (header_query.discretization == OutputTime::DiscreteSpace::NATIVE_DATA);
 		bool boundary;
 		if (is_native || this->interpolation_==DataInterpolation::identic_msh || this->interpolation_==DataInterpolation::equivalent_msh) {
-			n_entities = dh_->mesh()->n_elements();
+			n_entities = ReaderCache::get_mesh(reader_file_)->n_elements(this->boundary_domain_); //dh_->mesh()->n_elements();
```
Comment (jbrezmorf) [inline src/fields/field_fe.cc:438]: set `n_entities` before the condition as it seems common to both branches. Possibly same holds for the boundary flag.

File: src/fields/field_fe.cc [hunk 6]
```diff
@@ -692,10 +687,10 @@ void FieldFE<spacedim, Value>::calculate_native_values(ElementDataCache<double>:
 	for (auto cell : dh_->own_range()) {
 		dof_size = cell.get_dof_indices(global_dof_indices);
 		LocDofVec loc_dofs = cell.get_loc_dof_indices();
-		data_vec_i = cell.elm_idx() * dof_size;
+		data_vec_i = source_target_mesh_elm_map_[cell.elm_idx()] * dof_size;
 		ASSERT_EQ_DBG(dof_size, loc_dofs.n_elem);
 		for (unsigned int i=0; i<dof_size; ++i, ++data_vec_i) {
-			(*data_vector)[ loc_dofs[i] ] += (*data_cache)[ global_dof_indices[i] ];
+			(*data_vector)[ loc_dofs[i] ] += (*data_cache)[ data_vec_i ];
 			++count_vector[ loc_dofs[i] ];
```
Comment (jbrezmorf) [inline src/fields/field_fe.cc:694]: Fine.

**Review by jbrezmorf — COMMENTED** [review]
Dobry, ale to check compatible mesh by chtelo trochu vice predelat. Ale nebude to snad slozite.

File: src/mesh/bc_mesh.cc [hunk 7]
```diff
@@ -75,8 +75,8 @@ const LongIdx *BCMesh::get_local_part() {
 }
 
 
-bool BCMesh::check_compatible_mesh( Mesh & mesh, vector<LongIdx> & bulk_elements_id, vector<LongIdx> & boundary_elements_id ) {
-	return parent_mesh_->check_compatible_mesh(mesh, bulk_elements_id, boundary_elements_id);
+bool BCMesh::check_compatible_mesh( Mesh & mesh, vector<LongIdx> & element_ids_map ) {
```
[the remark this answers is not in the document]
Comment (dflanderka) [reply src/mesh/bc_mesh.cc]: We can't try it now. All fields in equation are constructed under computational mesh. We need to change implementation when boundary fields will be constructed under BCMesh and this class will bu full implemented.

**Pull request merged**